core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -627,7 +627,16 @@ private[spark] class BlockManager(
   override def getLocalBlockData(blockId: BlockId): ManagedBuffer = {
     if (blockId.isShuffle) {
       logDebug(s"Getting local shuffle block ${blockId}")
-      shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      try {
+        shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      } catch {
+        case e: IOException =>
+          if (conf.get(config.STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
+            FallbackStorage.read(conf, blockId)
+          } else {
+            throw e
+          }
```

Comment:
Can you explain the reasoning here? So once any executor cannot get local shuffle block, it will always try to read from fallback storage?

Reply:
Yes. If `STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH` is defined, it tried to fallback. Currently, it happens when `FALLBACK_BLOCK_MANAGER_ID` has the blocks.

Reply:
Is there any check if `FALLBACK_BLOCK_MANAGER_ID` has the blocks before calling `FallbackStorage.read`? Looks like if `getBlockData` throws `IOException`, the block manager will just go to read from fallback storage.

Reply:
It's already `IOException` and `FallbackStorage.read` throws `IOException` for non-exist files and it's fine and legitimate in the `WorkerDecomission` context, @viirya . Please note the following.
- The whole Worker decommission (including `storage decommission`) is designed as a best-effort approach.
- It's because the main use case is K8s graceful shutdown with the default period (`30s`). We can increase it, but we cannot set it infinite value technically. It means executor hangs in case of disk full situation.
- What we are aiming is to rescue data as much as possible, but `100%` is not guaranteed always.
- Lastly, data block selection(shuffle or rdd) was a random-order from the beginning. It became worse when there are multiple shuffle across multiple executors.

Due to the above reasons, we introduced `SPARK-33387 Support ordered shuffle block migration` to rescue as complete as possible. However, we can still easily imagine that multiple shuffles coexist on an executor in a skewed manner. While the other executor succeeds to complete the migration, the skewed executor may fail to complete the migration in the given grace period.

Reply:
I think this would probably fail quickly if the external storage doesn’t have the block right? So no need to do a separate check to see if the block is present, the failure message is good enough.

Reply:
Yes, it does, @holdenk .

Reply:
Make sense. Thanks @holdenk